gravatar/src/main/java/com/gravatar/services/AvatarService.kt

```diff
@@ -22,7 +23,8 @@ public class AvatarService(private val okHttpClient: OkHttpClient? = null) {
         const val LOG_TAG = "AvatarService"
     }
 
-    private val coroutineScope = CoroutineScope(GravatarSdkDI.dispatcherDefault)
+    // Run onResponse and onError callbacks on the main thread
+    private val coroutineScope = CoroutineScope(Dispatchers.Main)
```

Comment:
Since it looks like I can't add review comments in code that was not changed I will add some stuff related to coroutine context switching here:

### Asynchronous work
I just noticed that `service.uploadImage` is not being called in a coroutine, which looks like an error, as it's probably the longest-running operation in this function. If that should indeed be running inside a coroutine, I would change this function to actually launch a coroutine ASAP, before line 38:

```kotlin
coroutineScope.launch {
    // content from lines 38-64 here (aka the whole function body)
}
```

### IO work
In this class it looks like the `service.uploadImage` call highly relies on network communication, so it's suited for running on the IO Dispatcher (a dispatcher that is optimized for disk/network operations). BTW 

So your code could have this just before line 43, closing the block after the function call, at line 64 (considering the code is now running inside a coroutine):
```kotlin
withContext(GravatarSdkDI.dispatcherIO) {
    service.uploadImage(...).enqueue(...)
}
```

### Main-thread work
Likewise, operations that should run in the Main thread should also use `withContext` where applicable around the blocks that make sense, such as callback calls, which are _usually_ expected to run in the main thread, especially if we don't have information about where the calling code is running.

Based on that, the code for the `onSuccess` should launch a coroutine in the Main dispatcher, so line 46 should be:
```kotlin
coroutineScope.launch(GravatarSdkDI.dispatcherMain) {
    // code here for checking success and calling the appropriate callback
}
```

### Something odd about the `onFailureCode`
I noticed that for the failure there's a `handleError` function that receives a `coroutineScope` as parameter, not sure exactly why. If the idea here is that `handleError` should _always_ be called from a coroutine, the best way you can do so is to convert it into a `suspend fun` and remove the `coroutineScope` argument and [the coroutine launch inside it](https://github.com/Automattic/Gravatar-SDK-Android/blob/d43fab05636329659b124b49f525b739200f415c/gravatar/src/main/java/com/gravatar/services/ErrorUtils.kt#L15), to enforce the caller having to launch a coroutine before calling `handleError`.

Then inside `handleError` you can change lines 15-17 to switch to the main context, so any code using this function will automatically use the correct Dispatcher for calling the `listener.onError`.

```kotlin
// inside ErrorUtils.kt, lines 15-17
withContext(GravatarSdkDI.dispatcherMain) {
    listener.onError(error)
}
```

You will also likely need to launch a coroutine around line 61:
```kotlin
coroutineScope.launch {
    // not really important to use the correct Dispatcher since the suspending handleError should take care
    handleError(t, gravatarUploadListener)
}
```

### More about using Dispatchers
You can find more about using Dispatchers and their differences [here](https://developer.android.com/kotlin/coroutines/coroutines-adv#main-safety).

Reply:
> In this class it looks like the service.uploadImage call highly relies on network communication, so it's suited for running on the IO Dispatcher (a dispatcher that is optimized for disk/network operations). BTW

We're using the Retrofit library, in this case the call to `enqueue()` will run the network request asynchronously and return results in the `Callback<ResponseBody>` block when ready. File reading is also done asynchronously (via the `RequestBody` object).

Reply:
Maybe you don't even need to launch a coroutine for the responses callbacks as they seem to be ran in the correct thread. Per the [documentation](https://square.github.io/retrofit/2.x/retrofit/retrofit2/Callback.html) from retrofit's `Callback` interface:

> Android: Callbacks are executed on the application's main (UI) thread.
JVM: Callbacks are executed on the background thread which performed the request.

Reply:
I didn't realize Retrofit was being used, my bad. In that case I agree with @eyedol, probably there's no need to launch a coroutine for any response callbacks (and maybe anything else in this class, since the thread management for the UI part will already be done by Retrofit).

Reply:
Great discussion here! 

I see two options:

- Let Retrofit handle the thread management using callbacks and `enqueue()` calls and remove everything related to Dispatchers because we won't need it anymore with that implementation. 
- Use sync Retrofit implementation with `execute()` and handle everything internaly in the GravatarSDK. For example:
```
CoroutineScope(GravatarSdkDI.dispatcherIO).launch {
    service.uploadImage("Bearer $accessToken", identity, filePart).execute().let { response ->
        if (response.isSuccessful) {
            withContext(GravatarSdkDI.dispatcherMain) {
                gravatarUploadListener.onSuccess(Unit)
            }
        } else {
            // Log the response body for debugging purposes if the response is not successful
            Logger.w(
                LOG_TAG,
                "Network call unsuccessful trying to upload Gravatar: $response.body",
            )
            withContext(GravatarSdkDI.dispatcherMain) {
                gravatarUploadListener.onError(errorTypeFromHttpCode(response.code()))
            }
        }
    }
}
```

I would go with the second option because it gives us control over where and how things are executed. Also, we probably need to use coroutines for more things than network requests, so I think we need to handle the Dispatcher injection anyway.

Reply:
Good point. Handling thread/dispatcher switching ourselves definitely gives us more flexibility and ways of optimizing the performance of some stuff. For instance, doing most processing in the `Default` dispatcher, network calls in `IO` dispatcher, and calling the callbacks in the `Main` dispatcher.

But if the idea is to use Coroutines for everything with Retrofit then I would suggest also using the coroutine-compatible Retrofit API. For instance, instead of having Services with functions that use `Call`s, replace them with `suspend` functions.

An example of how that would look like in the `GravatarApiService` (if I'm not mistaken):

```kotlin
// BEFORE
@Multipart
@POST("upload-image")
fun uploadImage(
    @Header("Authorization") authorization: String,
    @Part identity: MultipartBody.Part,
    @Part data: MultipartBody.Part,
): Call<ResponseBody>

// AFTER (with coroutines)
@Multipart
@POST("upload-image")
suspend fun uploadImage(
    @Header("Authorization") authorization: String,
    @Part identity: MultipartBody.Part,
    @Part data: MultipartBody.Part,
): Response<ResponseBody>
```

But now I think we might be going way beyond the scope of this PR and talking about general improvements to the codebase, which should be tackled separately.

Reply:
Great inputs @hamorillo and @thomashorta. Makes sense to go all in with using coroutines as it gives control and yes better if this is handled as a separate PR as it will require a bit of rework.

I don't mind taking it as my first real coding contribution though :)

Reply:
> I would suggest also using the coroutine-compatible Retrofit API. For instance, instead of having Services with functions that use Calls, replace them with suspend functions.

Yeah, it makes sense. In fact, as we are using the [OpenApi generator](https://github.com/Automattic/Gravatar-SDK-Android?tab=readme-ov-file#generating-api-code-from-openapi-definitions) is just a matter of changing one [configuration](https://openapi-generator.tech/docs/generators/kotlin#config-options).

<img width="910" alt="image" src="https://github.com/Automattic/Gravatar-SDK-Android/assets/6974554/4aa60334-11cf-44ae-8176-ad73157ced71">

> My main concern (and I might be being nitpicky) is about which Coroutine Context is being used, since the CoroutineScope is currently being created and managed by the class, without being attached to any lifecycle-aware component nor the calling code (e.g.: via a suspend function). 

I don't think you are nitpicking. That's an important point. We probably should declare our functions as `suspend`.

Anyway, I would use this PR to fix the issue detected in WPAndroid in a simple way (for example, by adding the Main dispatcher), and we can work on the refactor later.